### DISCUSSION TITLE
Best approach to union multiple queries

### DISCUSSION THREAD

**Discussion by toochevere** [ravendb/ravendb#15659] [Q&A] [answered]
What is the best / recommended approach to create a combined (union concept from SQL) query of multiple sub-queries?   I'm referring to the opposite functionality as is [provided by `Intersect`](https://ravendb.net/docs/article-page/5.4/csharp/indexes/querying/intersection).  In the case of `Intersect`, the queries are intersected on the server side and only the documents matching _all_ sub-queries are returned.   I want to do the opposite; return documents matching _any_ sub-query. 

*Example:*

```
IList<TShirt> results = session.Query<TShirts_ByManufacturerColorSizeAndReleaseYear.Result, TShirts_ByManufacturerColorSizeAndReleaseYear>()
       .Where(x => x.Manufacturer == "Raven")
       .Union()
       .Where(x => x.Color == "Blue" && x.Size == "Small")
       .Union()
       .Where(x => x.Color == "Gray" && x.Size == "Large")
       .OfType<TShirt>()
       .ToList();
```

The above query would return _all_ shirts manufactured by Raven, and all that are blue and small, and all that are gray and large.  

*Possible approaches:*

Two possible ways for how to do this, but each presents some down-sides for me.  

1. A multi-map index (those are so awesome) could be used to union results, but in this case I don't know the conditions ahead of time. 
2. A series of lazy queries could be used, which get sent all at once.  Then just union the results myself.  Will work but feels a bit clumsy.   In fact, I did dig up [an old answer](https://groups.google.com/g/ravendb/c/TRlFFdfyaGo) that points this direction. 

Is number 2 still the recommended way?  Would there be some benefit to formalize this into a method like `Union` in the model of `Intersect`? (arguably just sugar, but still it reads much nicer)

**Comment by vlko** [comment, marked as answer]
For that purpose we have PredicateBuilder extension

```
namespace System.Linq
{
    public static class PredicateBuilder
    {
        public static Expression<Func<T, bool>> NullExpression<T>() { return null; }
        public static Expression<Func<T, bool>> True<T>() { return f => true; }
        public static Expression<Func<T, bool>> False<T>() { return f => false; }

        public static Expression<Func<T, bool>> Or<T>(this Expression<Func<T, bool>> expr1,
                                                            Expression<Func<T, bool>> expr2)
        {
            if (expr1 != null)
            {
                if (expr2 != null)
                {
                    return Expression.Lambda<Func<T, bool>>
                        (Expression.OrElse(expr1.Body, expr2.Body), expr1.Parameters);
                }
                return expr1;
            }
            return expr2;
        }

        public static Expression<Func<T, bool>> And<T>(this Expression<Func<T, bool>> expr1,
                                                             Expression<Func<T, bool>> expr2)
        {
            if (expr1 != null)
            {
                if (expr2 != null)
                {
                    return Expression.Lambda<Func<T, bool>>
                        (Expression.AndAlso(expr1.Body, expr2.Body), expr1.Parameters);
                }
                return expr1;
            }
            return expr2;
        }
    }
}

```
We are using it for dynamic filters like this (District is list of some filter values):
```
if (model.District != null && model.District.Length > 0)
{
    var filterExpression = PredicateBuilder.NullExpression<Company>();
    foreach (var district in model.District)
    {
        var filter = district;
        filterExpression = filter == CompanyFacets.NullValueConst
            ? filterExpression.Or(item => item.District == null)
            : filterExpression.Or(item => item.District == filter);
    }
    searchQuery = searchQuery.Where(filterExpression);
}
```
you can build any complex filter with those extensions
